# Master's thesis in **Bioinformatics and Computational Biology**

This is the repository for my Master's Thesis in **Bioinformatics and Computational Biology** at the Universidad AutÃ³noma de Madrid (UAM), course 2024/25. It contains the code used. 

**ğŸ‘©â€ğŸ’» Author**: **Alba MÃ©ndez Alejandre**  

## ğŸ§¬ Brief description
This work applies variant calling from **scRNA-seq** to link genetic mutations to cellular phenotypes, using a **customized pipeline** on mouse and human esophageal data. It highlights the importance of experimental design and filtering for reliable mutation detection.

## ğŸ¯ Objectives
The main objective of this work is **to assess whether single-cell transcriptomics is suitable for reliable somatic variant detection in mouse esophageal epithelium, and to associate genotype-to-phenotype relations using single cell RNA-seq.**

- Apply and customize variant caller for scRNA-seq in mouse and human esophagus.
- Characterize the diversity of transcriptomic states in the mouse esophageal epithelium. 
- Map variants onto UMAP embeddings. 

## ğŸ› ï¸ Tools & Technologies used
- (variant caller)
- `Seurat`, `scVelo`, `Slingshot` (transcriptomic analysis & trajectory inference)
- `VEP` (variant annotation)

## ğŸ“Š Data
We obtained the HCA data from previous work (GonzÃ¡lez-MenÃ©ndez, 2024), originally taken from: 
> Madissoon, E., Wilbrey-Clark, A., Miragaia, R.J. et al. scRNA-seq assessment of the human lung, spleen, and esophagus tissue stability after cold preservation. Genome Biol 21, 1 (2020). https://doi.org/10.1186/s13059-019-1906-x

The mouse data was obtained from a publicly available dataset.

## ğŸ“ Repository structure
The repository holds independent scripts for each dataset: 
```
â”œâ”€â”€ docs/ # quarto book
â”‚
â”œâ”€â”€ Introduction.qmd # installations, brief explanation of configurations used
â”‚
â”œâ”€â”€ human/ # ğŸ§Human analysis
â”‚ â”œâ”€â”€ 1_Inspection.qmd # preliminary inspection of the dataset
â”‚ â”œâ”€â”€ 2_GeneExpression.qmd # calculate average gene expression for sets of genes
â”‚ â”œâ”€â”€ 3_VennDiagrams.qmd # obtain venn diagrams for sets of genes or mutated genes of interest
â”‚ â”œâ”€â”€ 4_UMAP_mapping.qmd # map mutated cells in the umap
â”‚ â””â”€â”€ mut_clones_analysis_hca.qmd # modifying seurat_obj@meta.data to add clones
â”‚
â”œâ”€â”€ mouse/ # ğŸ­ Mouse analysis
â”‚ â”œâ”€â”€ fastQC/
â”‚ â”‚   â”œâ”€â”€ run1/ # multiqc report for the first sequencing run
â”‚ â”‚   â””â”€â”€ run2/ # multiqc report for the second sequencing run
â”‚ â”œâ”€â”€ 1-4_merge_seurat_fixedrank.R # script to filter out droplets and doublets from the matrices
â”‚ â”œâ”€â”€ 1_DataProcessing.qmd # notebook which contains the bash scripts used to download the files, and oerform the alignment with STARsolo.
â”‚ â”œâ”€â”€ 2_ClusteringCellAnnotation.qmd # seurat pipeline to cluster the cells and annotate them
â”‚ â”œâ”€â”€ 3_Velocity_inference.qmd # infer velocity and pseudotime trajectory
â”‚ â”œâ”€â”€ 4_VariantCalling.qmd  # steps to perform variant calling and pre-filtering processing, as well as TSV to VCF conversion
â”‚ â”œâ”€â”€ 5_AnnotationVariants.qmd # exploration of the VEP output
â”‚ â””â”€â”€ 6_GOAnalysis.qmd # GO analysis of the mutated genes
â”‚
â””â”€â”€  scripts/ # other general scripts
  â””â”€â”€  annotate_vep.sh # automated annotation with vep

```