<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.55">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>3. SComatic calling – Master's thesis in Bioinformatics and Computational Biology</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<link href="../Mouse/5_AnnotationVariants.html" rel="next">
<link href="../Mouse/3_Velocity_inference.html" rel="prev">
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../Mouse/1_DataProcessing.html">Mouse analysis</a></li><li class="breadcrumb-item"><a href="../Mouse/4_VariantCalling.html"><span class="chapter-title">3. SComatic calling</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">Master’s thesis in Bioinformatics and Computational Biology</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/alba179/master-thesis-bioinfo" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
</div>
    </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">Mouse and Human SComatic analysis</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../intro.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">Introduction</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Mouse analysis</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../Mouse/1_DataProcessing.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">1. Data Pre-processing</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../Mouse/2_ClusteringCellAnnotation.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">2. Clustering and Cell annotation</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../Mouse/3_Velocity_inference.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">3. Inference of velocity</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../Mouse/4_VariantCalling.html" class="sidebar-item-text sidebar-link active"><span class="chapter-title">3. SComatic calling</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../Mouse/5_AnnotationVariants.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">5. Functional annotation of variants</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../Mouse/6_GOAnalysis.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">4. Functional Analysis</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="false">
 <span class="menu-text">Human HCA analysis</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../Human/3_VennDiagrams_hca.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">3 - Visualization of intersected genes with Venn diagrams in esophagus HCA dataset</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="4">
    <h2 id="toc-title">Index</h2>
   
  <ul>
  <li><a href="#script-to-call-somatic-variants" id="toc-script-to-call-somatic-variants" class="nav-link active" data-scroll-target="#script-to-call-somatic-variants">3.1. Script to call somatic variants</a></li>
  <li><a href="#singlecellgenotype-calling-to-obtain-mutated-cell-barcodes" id="toc-singlecellgenotype-calling-to-obtain-mutated-cell-barcodes" class="nav-link" data-scroll-target="#singlecellgenotype-calling-to-obtain-mutated-cell-barcodes">3.2. SingleCellGenotype calling to obtain mutated cell barcodes</a></li>
  <li><a href="#discard-variants-falling-in-repetitive-regions" id="toc-discard-variants-falling-in-repetitive-regions" class="nav-link" data-scroll-target="#discard-variants-falling-in-repetitive-regions">3.3. Discard variants falling in repetitive regions</a>
  <ul>
  <li><a href="#filter-scomatic-output-to-discard-variants-falling-in-repetitive-regions" id="toc-filter-scomatic-output-to-discard-variants-falling-in-repetitive-regions" class="nav-link" data-scroll-target="#filter-scomatic-output-to-discard-variants-falling-in-repetitive-regions">Filter SComatic output to discard variants falling in repetitive regions</a></li>
  </ul></li>
  <li><a href="#filtering-of-tsvs" id="toc-filtering-of-tsvs" class="nav-link" data-scroll-target="#filtering-of-tsvs">3.4. Filtering of TSVs</a>
  <ul>
  <li><a href="#filter-the-tsvs-with-only-pass-in-filter-column" id="toc-filter-the-tsvs-with-only-pass-in-filter-column" class="nav-link" data-scroll-target="#filter-the-tsvs-with-only-pass-in-filter-column">Filter the TSVs with only PASS in FILTER column</a></li>
  <li><a href="#filter-the-tsvs-with-pass-or-multiple_cell_types-in-filter-column" id="toc-filter-the-tsvs-with-pass-or-multiple_cell_types-in-filter-column" class="nav-link" data-scroll-target="#filter-the-tsvs-with-pass-or-multiple_cell_types-in-filter-column">Filter the TSVs with PASS or Multiple_cell_types in FILTER column</a></li>
  </ul></li>
  <li><a href="#convert-the-tsv-files-to-vcf" id="toc-convert-the-tsv-files-to-vcf" class="nav-link" data-scroll-target="#convert-the-tsv-files-to-vcf">3.5. Convert the TSV files to VCF</a>
  <ul>
  <li><a href="#check-vcf-files-integrity" id="toc-check-vcf-files-integrity" class="nav-link" data-scroll-target="#check-vcf-files-integrity">Check VCF files’ integrity</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../Mouse/1_DataProcessing.html">Mouse analysis</a></li><li class="breadcrumb-item"><a href="../Mouse/4_VariantCalling.html"><span class="chapter-title">3. SComatic calling</span></a></li></ol></nav>
<div class="quarto-title">
<h1 class="title"><span class="chapter-title">3. SComatic calling</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>We are going to perform somatic variant calling with the novel algorithm from Muyas et.al (2024) doi: https://doi.org/10.1038/s41587-023-01863-z</p>
<section id="script-to-call-somatic-variants" class="level2">
<h2 class="anchored" data-anchor-id="script-to-call-somatic-variants">3.1. Script to call somatic variants</h2>
<ul>
<li>min_ac_cells = 2 (value by default); in HCA was used 1</li>
<li>min_cell_types = 2</li>
<li>max_cell_types = 1</li>
</ul>
<pre class="{bash}"><code>#!/bin/bash
# Script with bams separated by run (run1 vs run2). Each run bams are analyzed in a scomatic pipeline. This means, the output of this script will give two final .tsv: one for the run of run1, and another one for the run of run2. Each of these runs/.tsv, will have 8 samples: sample1_Epi, samp2_Epi, samp3_Epi, samp4_Epi, samp5_Epi, samp6_Epi, samp7_Epi, samp8_Epi. 
# There are 6 sample which are control adult, + one sample is old CTL, and + one sample is old DEN. 
# Each of the sample has the key like "sample1_Epi", "samp2_Epi", etc. 
# DISCLAIMER: ALWAYS USE THE SAME REFERENCE GENOME!!! If you have aligned you rreads with mm10, scomatic will have to be ran with mm10. No coordinates' change with bedtools nor similar strange things. 
# SComatic nowadays has only panel of normales (PON) and editing sites (A-to-I events) for: GRCh38 and mm10. There aren't PON nor editing sites for mouse assembly mm39. 
# The PON for human has been made with GATK data 1000 genomes project. No info of how they did it.
# The PON for mouse mm10 was done in July 20, using mm10 Tabula muris data. No info of how exactly they did it. 

# Run this script in a environment with scomatic installed. 

SCOMATIC=~/bin/SComatic-main

run_names=("run1" "run2") # list of runs

echo "Starting Scomatic analysis."


i=0
for run in "${run_names[@]}";
do
    i=$((i + 1)) 

  # iterate though each bam in the directory where they are located
  bam_dir="/media/storage/mcGinn_2021/STARalignment/bam/${run}"
  output_dir="/media/storage/mcGinn_2021/scomatic/output/${run}"

  mkdir -p $output_dir

  logfile="$output_dir/scomatic_log.txt"
  mkdir -p "$(dirname "$logfile")"

  log_message() {
      local log_time
      log_time=$(date +"%Y-%m-%d %H:%M:%S")
      local message="[$log_time] $1"
      echo "$message" &gt;&gt; "$logfile"
      echo "$message"
  }

  # Initialize log file
  echo "### Scomatic Pipeline Log ###" &gt; "$logfile"
  log_message "Running scomatic pipeline for run: ${run}"


  # Step 1: Splitting alignment file in cell type specific bams
  log_message "Step 1: Splitting alignment file in cell type specific bams..."

  output_dir1=$output_dir/Step1_BamCellTypes
  mkdir -p $output_dir1
  meta_dir="/media/storage/mcGinn_2021/scomatic/markers"

  # Iterate through each BAM file in the bam_dir
  for bam_file in "$bam_dir"/*.bam
  do
      # Extract the sample name from the file name
      base_name=$(basename "$bam_file" Aligned.sortedByCoord.out.bam) # run1_lib1
      sample=${base_name#*_} # remove the prefix up to '_', so that we get only lib1

      echo "Started sample: $sample"

      meta_file="$meta_dir/${run}/esoph_markers_scomatic_${base_name}.tsv"
      
      # Run the Python script with the appropriate arguments
      python $SCOMATIC/scripts/SplitBam/SplitBamCellTypes.py --bam "$bam_file" \
          --meta "$meta_file" \
          --id "$sample" \
          --n_trim 5 \
          --max_nM 5 \
          --max_NH 1 \
          --outdir "$output_dir1"

      log_message "Finished sample: $sample"

      sample_list+=("$sample")
  done

  # Now, we have each .bam file for each sample located in =$output_dir/Step1_BamCellTypes. In total, 16 .tsv
  # The bam file will be like: run1_lib1.lib1_Epi.bam


  #######################################
  # Step 2: Collecting base count information
  
  REF=/media/storage/reference_genomes/Mus_musculus.GRCm38.dna.primary_assembly.fa

  output_dir2=$output_dir/Step2_BaseCellCounts
  mkdir -p $output_dir2

  for bam in "$output_dir1"/*.bam
  do
    
    # Cell type, which will be lib1 in this case
    cell_type=$(basename $bam | awk -F'.' '{print $(NF-1)}') # this takes the cell_type from the bam name (the second field to last by separating by '.')

    # Temp folder
    temp=$output_dir2/temp_${cell_type}
    mkdir -p $temp

    log_message "Processing base counts for cell type: $cell_type"

    # Command line to submit to cluster
    python $SCOMATIC/scripts/BaseCellCounter/BaseCellCounter.py --bam $bam \
      --ref $REF \
      --chrom all \
      --out_folder $output_dir2 \
      --min_bq 30 \
      --tmp_dir $temp \
      --nprocs 10

    rm -rf $temp

    log_message "Finished base count processing for cell type: $cell_type"
  done

  log_message "Listing files in $output_dir2 before renaming:"
  ls "$output_dir2"
  # Now, we have each .tsv for each sample in output_dir2=$output_dir/Step2_BaseCellCounts
  # We are going to move both .tsv from the same sample to a new folder inside output_dir2, named after the sample name.

  # echo "Organizing .tsv files into sample-specific folders..."
  log_message "Renaming .tsv files..."


  for tsv in "$output_dir2"/*.tsv
  do
      # Create a new directory for the sample
      # Move the .tsv files corresponding to the sample into the sample directory
 
      base_name=$(basename "$tsv")
      
      # lib1.lib1_Epi.tsv ; we want lib1.tsv -&gt; this will be our unique "cell types"
    
      # Extract run name, and spec cell
      sample_run=$(echo "$base_name" | awk -F'.' '{print $1}') # extracts lib1
      spec_cell=$(echo "$base_name" | awk -F'_' '{print $2}') # extracts Epi.tsv

      new_filename="${output_dir2}/${sample_run}_${spec_cell}" # lib1_Epi.tsv

      # Change each name of the .tsv files by adding its run
      # mv "$output_dir2/${sample_name}"*.tsv "$sample_dir/${run_name}${sample_name}"*.tsv
      mv "$tsv" "$new_filename"

      # echo "Moved .tsv files for sample: $sample to $sample_dir"
      log_message "Renamed $tsv to $new_filename"
  done

  log_message "Listing files in $output_dir2 after renaming:"
  ls "$output_dir2"

  # echo "Moved the files successfully!"
  log_message "Renamed the files successfully!"

  #######################################
  # Step 3: Merging base count matrices
  log_message "Step 3: Merging base count matrices..."

  output_dir3=$output_dir/Step3_BaseCellCountsMerged
  mkdir -p $output_dir3

  log_message "Merging base count matrices into a single .tsv file..."

  python $SCOMATIC/scripts/MergeCounts/MergeBaseCellCounts.py \
    --tsv_folder ${output_dir2} \
    --outfile ${output_dir3}/${run}.BaseCellCounts.AllCellTypes.tsv


  #######################################
  # Step 4: Detection of somatic mutations
  log_message "Step 4: Detection of somatic mutations..."

  # Step 4.1
  log_message "Step 4.1: Variant calling..."
  output_dir4=$output_dir/Step4_VariantCalling
  mkdir -p $output_dir4

  python $SCOMATIC/scripts/BaseCellCalling/BaseCellCalling.step1.py \
            --infile ${output_dir3}/${run}.BaseCellCounts.AllCellTypes.tsv \
            --outfile ${output_dir4}/${run} \
        --max_cell_types 1 \
        --min_cell_types 2 \
            --ref $REF


  # Step 4.2
  log_message "Step 4.2: Somatic mutation detection..."
  editing=$SCOMATIC/RNAediting/AllEditingSites.mm10.txt
  PON=$SCOMATIC/PoNs/PoN.scRNAseq.mm10.tsv

  python $SCOMATIC/scripts/BaseCellCalling/BaseCellCalling.step2.py \
            --infile ${output_dir4}/${run}.calling.step1.tsv \
            --outfile ${output_dir4}/${run}.calling.step2.tsv \
            --editing $editing \
            --pon $PON

  # extra step: Intersection with bed file
  log_message "Extra step: Intersection with bed file..."
  bedtools intersect -header -a ${output_dir4}/${run}.calling.step2.tsv -b $SCOMATIC/bed_files_of_interest/UCSC.mm10.without.repeatmasker.bed | awk '$1 ~ /^#/ || $6 == "PASS"' &gt; ${output_dir4}/${run}.calling.step2.pass.tsv

  log_message "Finished scomatic pipeline for ${run} successfully!"

done

log_message "Pipeline completed successfully for both runs!"

# Display log file
echo "Log file saved to: $logfile"
cat "$logfile"

</code></pre>
<p>Execute:</p>
<pre class="{bash}"><code># [matterhorn]
./SComatic_mouse.sh
</code></pre>
</section>
<section id="singlecellgenotype-calling-to-obtain-mutated-cell-barcodes" class="level2">
<h2 class="anchored" data-anchor-id="singlecellgenotype-calling-to-obtain-mutated-cell-barcodes">3.2. SingleCellGenotype calling to obtain mutated cell barcodes</h2>
<pre class="{bash}"><code>#!/bin/bash
# SComatic extra functionality: Computing the genotype for each cell at the variant sites
# 01/10/2024
# Author: Alba Méndez Alejandre
# This is going to allow us to map each variant in the UMAP

SCOMATIC=~/bin/SComatic-main

run_names=("run1" "run1") # list of runs

REF=/media/storage/reference_genomes/Mus_musculus.GRCm38.dna.primary_assembly.fa
meta_dir=/media/storage/mcGinn_2021/scomatic/markers

i=0
for run in "${run_names[@]}";
do
    i=$((i + 1))

    output_dir="/media/storage/mcGinn_2021/scomatic/output/${run}"
    output_dir1=$output_dir/Step1_BamCellTypes
    output_dir4=$output_dir/Step4_VariantCalling

    STEP4_2_pass=${output_dir4}/${run}_modif.calling.step2.pass.tsv # modified chr1 to 1 tsv file

    output_dir7=$output_dir/SingleCellAlleles
    mkdir -p $output_dir7
    
    
    for bam in $(ls -d $output_dir1/*bam);do  
            cell_type=$(basename $bam | awk -F'.' '{print $(NF-1)}')
    
        temp=$output_dir7/temp_${cell_type}
            mkdir -p $temp

        meta_file="$meta_dir/${run}/esoph_markers_scomatic_${run}_${cell_type}.tsv"
        
            python $SCOMATIC/scripts/SingleCellGenotype/SingleCellGenotype.py --bam $bam  \
                --infile ${STEP4_2_pass}   \
                --nprocs 10   \
                --meta $meta_file   \
                --outfile ${output_dir7}/${run}_${cell_type}.single_cell_genotype.tsv \
                --tmp_dir $temp  \
                --ref $REF \
            --alt_flag All
        
        echo "bam file is: $bam, Infile is: ${STEP4_2_pass}, meta file is $meta_file and outfile is ${output_dir7}/${cell_type}.single_cell_genotype.tsv"

            rm -rf $temp
        done
done</code></pre>
<pre class="{bash}"><code># [matterhorn]
cd /home/alba/scripts/scomatic

./variants_mapping.sh</code></pre>
</section>
<section id="discard-variants-falling-in-repetitive-regions" class="level2">
<h2 class="anchored" data-anchor-id="discard-variants-falling-in-repetitive-regions">3.3. Discard variants falling in repetitive regions</h2>
<p>Creation of bed file with regions overlapping with repetitive ones, in order to remove variants falling in repetitive regions.</p>
<pre class="{bash}"><code>bedtools intersect -v -a mm10_genes.bed -b mm10_RepeatMasker.bed &gt; UCSC.m10.without.repeatmasker.bed
</code></pre>
<section id="filter-scomatic-output-to-discard-variants-falling-in-repetitive-regions" class="level3">
<h3 class="anchored" data-anchor-id="filter-scomatic-output-to-discard-variants-falling-in-repetitive-regions">Filter SComatic output to discard variants falling in repetitive regions</h3>
<pre class="{bash}"><code>conda activate SComatic
SCOMATIC=~/bin/SComatic-main
output_dir4=/media/storage/mcGinn_2021/scomatic/output/run1/Step4_VariantCalling

bedtools intersect -header -a ${output_dir4}/run1.calling.step2.tsv -b $SCOMATIC/bed_files_of_interest/UCSC.mm10.without.repeatmasker.bed | awk '$1 ~ /^#/ || $6 == "PASS"' &gt; ${output_dir4}/run1.calling.step2.pass.tsv

output_dir4=/media/storage/mcGinn_2021/scomatic/output/run2/Step4_VariantCalling

bedtools intersect -header -a ${output_dir4}/run2.calling.step2.tsv -b $SCOMATIC/bed_files_of_interest/UCSC.mm10.without.repeatmasker.bed | awk '$1 ~ /^#/ || $6 == "PASS"' &gt; ${output_dir4}/run2.calling.step2.pass.tsv</code></pre>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note:
</div>
</div>
<div class="callout-body-container callout-body">
<p>*calling.step2.pass.tsv files contain only FILTER = PASS variants that don’t fall in repetitive regions</p>
</div>
</div>
</section>
</section>
<section id="filtering-of-tsvs" class="level2">
<h2 class="anchored" data-anchor-id="filtering-of-tsvs">3.4. Filtering of TSVs</h2>
<section id="filter-the-tsvs-with-only-pass-in-filter-column" class="level3">
<h3 class="anchored" data-anchor-id="filter-the-tsvs-with-only-pass-in-filter-column">Filter the TSVs with only PASS in FILTER column</h3>
<pre class="{bash}"><code>
seq_ids=("run1" "run2")
wd="/media/storage/mcGinn_2021/scomatic/output/"

for seq in "${seq_ids[@]}"; do
   
    awk 'BEGIN {FS="\t"; OFS="\t"} NR &lt;= 29 || ($6 == "PASS" &amp;&amp; !/^#/) {print}' "${wd}/${seq}/Step4_VariantCalling/${seq}.calling.step2.tsv" &gt; "${wd}/${seq}/Step4_VariantCalling/${seq}_filtered.calling.step2.tsv"

done
</code></pre>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note:
</div>
</div>
<div class="callout-body-container callout-body">
<p>*_filtered.calling.step2.tsv files contain only FILTER = PASS variants, including variants that fall in repetitive regions.</p>
</div>
</div>
</section>
<section id="filter-the-tsvs-with-pass-or-multiple_cell_types-in-filter-column" class="level3">
<h3 class="anchored" data-anchor-id="filter-the-tsvs-with-pass-or-multiple_cell_types-in-filter-column">Filter the TSVs with PASS or Multiple_cell_types in FILTER column</h3>
<p>In order to analyze the variants that are arked solely as “Multiple_cell_types”, we obtain them via an awk command.</p>
<pre class="{bash}"><code>
seq_ids=("run1" "run2")
wd="/media/storage/mcGinn_2021/scomatic/output/"

for seq in "${seq_ids[@]}"; do

    awk 'BEGIN {FS="\t"; OFS="\t"} NR &lt;= 29 || ($6 == "PASS" || $6 == "Multiple_cell_types") &amp;&amp; !/^#/ {print}' "${wd}/${seq}/Step4_VariantCalling/${seq}.calling.step2.tsv" &gt; "${wd}/${seq}/Step4_VariantCalling/${seq}_mult.calling.step2.tsv"
done
</code></pre>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note:
</div>
</div>
<div class="callout-body-container callout-body">
<p>*_mult.calling.step2.tsv files contain only FILTER = PASS|Multiple_cell_types variants, including variants that fall in repetitive regions.</p>
</div>
</div>
<pre class="{bash}"><code># [matterhorn]
conda activate SComatic
SCOMATIC=~/bin/SComatic-main
output_dir4=/media/storage/mcGinn_2021/scomatic/output/run1/Step4_VariantCalling

# for multiple_cell type
bedtools intersect -header -a ${output_dir4}/run1_mult.calling.step2.tsv -b $SCOMATIC/bed_files_of_interest/UCSC.mm10.withouvt.repeatmasker.bed | awk '$1 ~ /^#/ || $6 == "PASS"' &gt; ${output_dir4}/run1_mult.calling.step2.pass.tsv

output_dir4=/media/storage/mcGinn_2021/scomatic/output/run2/Step4_VariantCalling

# for multiple_cell type
bedtools intersect -header -a ${output_dir4}/run2_mult.calling.step2.tsv -b $SCOMATIC/bed_files_of_interest/UCSC.mm10.without.repeatmasker.bed | awk '$1 ~ /^#/ || $6 == "PASS"' &gt; ${output_dir4}/run2_mult.calling.step2.pass.tsv
</code></pre>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note:
</div>
</div>
<div class="callout-body-container callout-body">
<p>*_mult.calling.step2.pass.tsv files contain only FILTER = PASS|Multiple_cell_types variants that don’t fall in repetitive regions</p>
</div>
</div>
</section>
</section>
<section id="convert-the-tsv-files-to-vcf" class="level2">
<h2 class="anchored" data-anchor-id="convert-the-tsv-files-to-vcf">3.5. Convert the TSV files to VCF</h2>
<p>The program outputs its files in a custom TSV format which makes it very complex to analyze in standard pipelines from variant calling. Thus, we are going to convert it to the standard format file for variant calling in bioinformatics, the VCFv4.3 (https://samtools.github.io/hts-specs/VCFv4.3.pdf). For that, we generated a custom script “3-5_TSVtoVCF.py”.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
What is a VCF?
</div>
</div>
<div class="callout-body-container callout-body">
<p>VCF is a text file format (most likely stored in a compressed manner). It contains meta-information lines (prefixed with “##”), a header line (prefixed with “#”), and data lines each containing information about a position in the genome and genotype information on samples for each position (text fields separated by tabs). Zero length fields are not allowed, a dot (“.”) must be used instead. In order to ensure interoperability across platforms, VCF compliant implementations must support both LF (“”) and CR+LF (“”) newline conventions.</p>
</div>
</div>
<pre class="{bash}"><code>cd /media/storage/mcGinn_2021/scomatic
conda activate d_vep

python 3-5_TSVtoVCF.py --add-info --add-celltypes output/run1/Step4_VariantCalling/run1.calling.step2.tsv output/run1/Step4_VariantCalling/run1.calling.step2.vcf
python 3-5_TSVtoVCF.py --add-info --add-celltypes output/run2/Step4_VariantCalling/run2.calling.step2.tsv output/run2/Step4_VariantCalling/run2.calling.step2.vcf</code></pre>
<section id="check-vcf-files-integrity" class="level3">
<h3 class="anchored" data-anchor-id="check-vcf-files-integrity">Check VCF files’ integrity</h3>
<p>We created a VCF file from the filtered SComatic output. Then, we confirmed that the file is properly formatted using both GATK4 and the EBI VCF validator. However, neither tool currently supports VCF version 4.4, they only support up to version 4.3.</p>
<pre class="{bash}"><code># [matterhorn]
conda activate samtools
samtools faidx Mus_musculus.GRCm39.dna.primary_assembly.faidx

conda activate d_gatk

gatk CreateSequenceDictionary -R Mus_musculus.GRCm39.dna.primary_assembly.fa 

REF=~/bin/SComatic-main/reference_genomes/Mus_musculus.GRCm38.dna.primary_assembly.fa 

gatk ValidateVariants -V output/run1/Step4_VariantCalling/run1_filtered.calling.step2.vcf -R $REF --validation-type-to-exclude ALLELES
gatk ValidateVariants -V output/run2/Step4_VariantCalling/run2_filtered.calling.step2.vcf -R $REF --validation-type-to-exclude ALLELES

vcf_validator -i output/run1/Step4_VariantCalling/run1_filtered.calling.step2.vcf -r summary,text,database
vcf_validator -i output/run2/Step4_VariantCalling/run2_filtered.calling.step2.vcf -r summary,text,database</code></pre>
<p>Both files are OK!!!</p>
<p>📁 Generated files:</p>
<ul>
<li>gatk ValidateVariants
<ul>
<li>Standard output message</li>
</ul></li>
<li>vcf_validator
<ul>
<li>Standard output message</li>
</ul></li>
</ul>
<p>It seems that both tools say the converted files are OK, as long as the files are marked as VCFv4.3, because they both don’t support VCFv4.4.</p>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../Mouse/3_Velocity_inference.html" class="pagination-link" aria-label="3. Inference of velocity">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-title">3. Inference of velocity</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../Mouse/5_AnnotationVariants.html" class="pagination-link" aria-label="5. Functional annotation of variants">
        <span class="nav-page-text"><span class="chapter-title">5. Functional annotation of variants</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




</body></html>